write-fonts/src/tables/gvar.rs

```diff
@@ -0,0 +1,451 @@
+        assert!(self.tuple_variation_headers.len() <= 4095);
+        let mut bits = self.tuple_variation_headers.len() as u16;
+        if self.shared_point_numbers.is_some() {
+            bits |= TupleVariationCount::SHARED_POINT_NUMBERS;
+        }
+        TupleVariationCount::from_bits(bits)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.tuple_variation_headers.is_empty()
+    }
+
+    fn compute_data_offset(&self) -> u16 {
+        let header_len = self
+            .tuple_variation_headers
+            .iter()
+            .fold(0usize, |acc, header| {
+                acc.checked_add(header.compute_size() as usize).unwrap()
+            });
+        (header_len + TupleVariationCount::RAW_BYTE_LEN + u16::RAW_BYTE_LEN)
+            .try_into()
+            .unwrap()
+    }
+
+    fn compute_size(&self) -> u32 {
+        if self.is_empty() {
+            return 0;
+        }
+
+        let data_start = self.compute_data_offset() as u32;
+        let shared_point_len = self
+            .shared_point_numbers
+            .as_ref()
+            .map(|pts| pts.compute_size())
+            .unwrap_or_default() as u32;
+        let tuple_data_len = self
+            .per_tuple_data
+            .iter()
+            .fold(0u32, |acc, tup| acc + tup.compute_size() as u32);
+        data_start + shared_point_len + tuple_data_len
+    }
+}
+
+impl Validate for GlyphVariationData {
+    fn validate_impl(&self, ctx: &mut ValidationCtx) {
+        const MAX_TUPLE_VARIATIONS: usize = 4095;
+        if !(0..=MAX_TUPLE_VARIATIONS).contains(&self.tuple_variation_headers.len()) {
+            ctx.in_field("tuple_variation_headers", |ctx| {
+                ctx.report("expected 0-4095 tuple variation tables")
+            })
+        }
+    }
+}
+
+impl FontWrite for GlyphVariationData {
+    fn write_into(&self, writer: &mut TableWriter) {
+        self.compute_tuple_variation_count().write_into(writer);
+        self.compute_data_offset().write_into(writer);
+        self.tuple_variation_headers.write_into(writer);
+        self.shared_point_numbers.write_into(writer);
+        self.per_tuple_data.write_into(writer);
+    }
+}
+
+impl HasLen for SharedTuples {
+    fn len(&self) -> usize {
+        self.tuples.len()
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct GvarPerTupleData {
+    private_point_numbers: Option<PackedPointNumbers>,
+    x_deltas: PackedDeltas,
+    y_deltas: PackedDeltas,
+}
+
+impl FontWrite for GvarPerTupleData {
+    fn write_into(&self, writer: &mut TableWriter) {
+        if let Some(points) = &self.private_point_numbers {
+            points.write_into(writer);
+        }
+        self.x_deltas.write_into(writer);
+        self.y_deltas.write_into(writer);
+    }
+}
+
+impl FontWrite for TupleVariationCount {
+    fn write_into(&self, writer: &mut TableWriter) {
+        self.bits().write_into(writer)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
```

Comment:
for such an important part that involves so many hand-coded bits the testing feels ... a little light?

Reply:
Definitely. My intention was to get a working version of this done, so that we could start playing around with it. I'm happy to try and come up with some more thorough integration test as followup.